src/library/scala/collection/mutable/HashMap.scala

```diff
@@ -107,6 +107,55 @@ class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
     }
   }
 
+  // Override updateWith for performance, so we can do the update while hashing
+  // the input key only once and performing one lookup into the hash table
+  override def updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = {
+    val hash = computeHash(key)
+    val indexedHash = index(hash)
+
+    var foundNode: Node[K, V] = null
+    var previousNode: Node[K, V] = null
+    table(indexedHash) match {
+      case null =>
+      case nd =>
+        @tailrec
+        def findNode(prev: Node[K, V], nd: Node[K, V], k: K, h: Int): Unit = {
+          if(h == nd.hash && k == nd.key) {
+            previousNode = prev
+            foundNode = nd
+          }
+          else if((nd.next eq null) || (nd.hash > h)) ()
+          else findNode(nd, nd.next, k, h)
+        }
+
+        findNode(null, nd, key, hash)
+    }
+
+    val previousValue = foundNode match {
+      case null => None
+      case nd => Some(nd.value)
+    }
+
+    val nextValue = remappingFunction(previousValue)
+
+    (previousValue, nextValue) match {
```

Comment:
We could avoid allocation here, maybe it doesn't matter though.

Reply:
Yeah I'm assuming either the Scalac backend optimizer of the JVM optimizer will be clever enough to elide the tuple here